src/cargo/core/resolver/features.rs

```diff
@@ -0,0 +1,504 @@
+                        opts.ignore_inactive_targets = true;
+                    }
+                    "compare" => opts.compare = true,
+                    "ws" => unimplemented!(),
+                    "host" => unimplemented!(),
+                    s => anyhow::bail!("-Zfeatures flag `{}` is not supported", s),
+                }
+            }
+            Ok(())
+        };
+        if let Some(feat_opts) = unstable_flags.features.as_ref() {
+            enable(feat_opts)?;
+        }
+        // This env var is intended for testing only.
+        if let Ok(env_opts) = std::env::var("__CARGO_FORCE_NEW_FEATURES") {
+            if env_opts == "1" {
+                opts.new_resolver = true;
+            } else {
+                let env_opts = env_opts.split(',').map(|s| s.to_string()).collect();
+                enable(&env_opts)?;
+            }
+        }
+        if has_dev_units {
+            // Decoupling of dev deps is not allowed if any test/bench/example
+            // is being built. It may be possible to relax this in the future,
+            // but it will require significant changes to how unit
+            // dependencies are computed, and can result in longer build times
+            // with `cargo test` because the lib may need to be built 3 times
+            // instead of twice.
+            opts.decouple_dev_deps = false;
+        }
+        Ok(opts)
+    }
+}
+
+/// Features flags requested for a package.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct RequestedFeatures {
+    pub features: FeaturesSet,
+    pub all_features: bool,
+    pub uses_default_features: bool,
+}
+
+impl RequestedFeatures {
+    /// Creates a new RequestedFeatures from the given command-line flags.
+    pub fn from_command_line(
+        features: &[String],
+        all_features: bool,
+        uses_default_features: bool,
+    ) -> RequestedFeatures {
+        RequestedFeatures {
+            features: Rc::new(RequestedFeatures::split_features(features)),
+            all_features,
+            uses_default_features,
+        }
+    }
+
+    /// Creates a new RequestedFeatures with the given `all_features` setting.
+    pub fn new_all(all_features: bool) -> RequestedFeatures {
+        RequestedFeatures {
+            features: Rc::new(BTreeSet::new()),
+            all_features,
+            uses_default_features: true,
+        }
+    }
+
+    fn split_features(features: &[String]) -> BTreeSet<InternedString> {
+        features
+            .iter()
+            .flat_map(|s| s.split_whitespace())
+            .flat_map(|s| s.split(','))
+            .filter(|s| !s.is_empty())
+            .map(InternedString::new)
+            .collect::<BTreeSet<InternedString>>()
+    }
+}
+
+impl ResolvedFeatures {
+    /// Returns the list of features that are enabled for the given package.
+    pub fn activated_features(&self, pkg_id: PackageId, is_build: bool) -> Vec<InternedString> {
```

Comment:
Similar to a different comment elsewhere, could `is_build` be replaced with something more desciptive? I'm imagining something like two methods (`activated_features` and `activated_features_for_build_dep`) or perhaps an `enum` with named variants here.